File: ios/Shared/Sources/Features/LmsCourseSummary.swift
import Foundation
import MoodleNativeCore
import MoodleNativeNetworking

public struct LmsCourseSummary: Codable, Identifiable, Equatable, Sendable {
    public let id: Int
    public let title: String
    public let courseCode: String?
    public let shortName: String
    public let summary: String?
    public let courseImageURL: URL?
    public let progress: Double?
    public let isFavorite: Bool
    public let academicSemester: AcademicSemester?
    public let scheduleSlots: [TimetableScheduleSlot]
    public let assignmentCount: Int
    public let unreadAnnouncementCount: Int
    public let unreadForumPostCount: Int
    public let isRecentlyAccessed: Bool
    public let detailURL: URL?

    public init(
        id: Int,
        title: String,
        courseCode: String?,
        shortName: String,
        summary: String?,
        courseImageURL: URL?,
        progress: Double?,
        isFavorite: Bool,
        academicSemester: AcademicSemester?,
        scheduleSlots: [TimetableScheduleSlot],
        assignmentCount: Int,
        unreadAnnouncementCount: Int,
        unreadForumPostCount: Int,
        isRecentlyAccessed: Bool,
        detailURL: URL?
    ) {
        self.id = id
        self.title = title
        self.courseCode = courseCode
        self.shortName = shortName
        self.summary = summary
        self.courseImageURL = courseImageURL
        self.progress = progress
        self.isFavorite = isFavorite
        self.academicSemester = academicSemester
        self.scheduleSlots = scheduleSlots
        self.assignmentCount = assignmentCount
        self.unreadAnnouncementCount = unreadAnnouncementCount
        self.unreadForumPostCount = unreadForumPostCount
        self.isRecentlyAccessed = isRecentlyAccessed
        self.detailURL = detailURL
    }

    public var progressFraction: Double? {
        guard let progress else {
            return nil
        }

        return max(0, min(progress, 100)) / 100
    }
}

File: android/features/src/main/kotlin/dev/example/moodlenative/features/LmsCourseSummary.kt
package dev.example.moodlenative.features

import dev.example.moodlenative.core.AcademicSemester
import dev.example.moodlenative.core.TimetableScheduleSlot
import java.net.URI

data class LmsCourseSummary(
    val id: Int,
    val title: String,
    val courseCode: String?,
    val shortName: String,
    val summary: String?,
    val courseImageURL: URI?,
    val progress: Double?,
    val isFavorite: Boolean,
    val academicSemester: AcademicSemester?,
    val scheduleSlots: List<TimetableScheduleSlot>,
    val assignmentCount: Int,
    val unreadAnnouncementCount: Int,
    val unreadForumPostCount: Int,
    val isRecentlyAccessed: Boolean,
    val detailURL: URI?,
) {
    val progressFraction: Double?
        get() = progress?.coerceIn(0.0, 100.0)?.div(100.0)
}
